spec.md

```diff
@@ -0,0 +1,269 @@
+
+```
+    InstructionStack ::= ".Stack"
+                       | InstructionStack Instruction
+```
+
+### The EEI
+
+The EEI contains fields relevant to VMs used for executing contracts on the Ethereum blockchain.
+
+**TODO**: In the definition of `EEI.useGas`, the field `gas` is treated like an `i64.const`, but here it's an `Int256`.
+          How do we reconcile these?
+          In EVM, it's an `Int256` in general.
+
+```
+    EEI ::= { statusCode : StatusCode
+            , accounts   : Accounts
+            , gas        : Int256
+            }
+```
+
+The `StatusCode` is defined by the [EVM-C API Standard](https://github.com/ethereum/evmc).
+See the repository for an explanation of each status code.
+
+```
+    StatusCode ::= "EVM_SUCCESS"
+                 | "EVM_FAILURE"
+                 | "EVM_INVALID_INSTRUCTION"
+                 | "EVM_UNDEFINED_INSTRUCTION"
+                 | "EVM_OUT_OF_GAS"
+                 | "EVM_BAD_JUMP_DESTINATION"
+                 | "EVM_STACK_OVERFLOW"
+                 | "EVM_STACK_UNDERFLOW"
+                 | "EVM_INVALID_MEMORY_ACCESS"
+                 | "EVM_REVERT"
+                 | "EVM_STATIC_MODE_ERROR"
+                 | "EVM_REJECTED"
+                 | "EVM_INTERNAL_ERROR"
+```
+
+Accounts store the four relevant fields for accounts on the Ethereum network.
+
+```
+    AccountId ::= Int160
+    Accounts  ::= Map{AccountId, Account}
+    Account   ::= { balance : Int256
+                  , code    : ByteString
+                  , storage : Map{Int256, Int256}
+                  , nonce   : Int256
+                  }
+```
+
+Initialisation of a contract
+----------------------------
+
+**TODO**
+
+Execution of the contract entry point
+-------------------------------------
+
+**TODO**
+
+Host functions available to the contract
+----------------------------------------
+
+The EEI specifies several host functions available to the user.
+Conceptually, it is defined as a wasm module which is available while executing ewasm contracts.
+The definition of the functions is *opaque*, meaning they are provided in terms of their update to the state and not in terms of wasm code.
+In the future, they may become more transparent as suitable wasm code is found to represent them.
+
+### EEI Module
+
+The function signatures are provided here in wast format:
+
+```wast
+    (module
+        (func (export "EEI.useGas")         (param i64)                 (result))
+        (func (export "EEI.call")           (param i64 i32 i32 i32 i32) (result i32))
+        (func (export "EEI.returnDataCopy") (param i32 i32 i32)         (result))
+    )
+```
+
+### `(call EEI.useGas (i64.const g))`
+
+Function to deduct the correct amount of gas for execution of an ewasm contract.
+
+1.  Load the value `i64.const G` from `EEI.gas`.
+
+2.  If `g <= G`, then:
+
+    i.  Set `EEI.gas` to `G - g`.
+
+3.  Else:
+
+    ii. Set `EEI.statusCode` to `EVM_OUT_OF_GAS` and Trap.
+
+```
```

Comment:
There's no transition leading into this snippet so I'm a bit confused about what's going on. Is what follows a formal specification of what you just laid out in English?

I also don't understand what `=>` means. Where is the `else` case?

Reply:
> There's no transition leading into this snippet so I'm a bit confused about what's going on. Is what follows a formal specification of what you just laid out in English?

Yep, it would be nice if there were clarification like "after prose snippet always goes snippet with formal spec"

> I also don't understand what `=>` means. Where is the else case?

You can see it such as: there are two rules, the first rule will be applied if `g <= G` and the second one if `g > G`. 

But I'm wondering, is there any exhaustivness testing? What happens if no rules couldn't be matched?

Reply:
I believe that `_=>_` and `_=>_if_` are explained in the `Notation` section at the top of the document, is it clear enough?

We could also make some sort of `if_then_else_` or `match_case_->_...` syntax for ensuring exhaustiveness, but there is not automatic check.

Note that this is not valid K code, I can make it valid K code if you prefer (it's very close).

Reply:
Oh nice to know! I assumed that this is a valid K!